### ISSUE TITLE
Modstrings

### ISSUE THREAD

**Issue by FelixErnst** [Bioconductor/Contributions#962]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/FelixErnst/Modstrings

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by FelixErnst** [comment]
I would like to suggest @hpages as a reviewer, since it is heavily intertwined with the Biostrings package.

**Comment by FelixErnst** [comment]
@hpages The error is specific to tokay2. Apparently there is a connection issue, when checking for  deprecated package usage by BiocCheck.

**Comment by FelixErnst** [comment]
Hi @hpages 

do you have any news on the contribution? Thanks for any update.

Best

**Comment by FelixErnst** [comment]
Hi Herve @hpages,

I would like to ask, whether you have some feedback on the package. Thanks for any update.

Best,
Felix

**Comment by mtmorgan** [comment]
@hpages is on a short vacation; i'll review this.

**Comment by FelixErnst** [comment]
Hi @mtmorgan, any news on the review? If this needs a bit of discussion before a first review, I am up for it.

**Comment by mtmorgan** [comment]
# 28 February, 2019

Sorry for the very slow review.

Generally the description, vignette, and man pages are excellent.

I am generally concerned with the use of unexported (accessible via `:::`) functions, C (via `.Call()`) calls, and direct slot access (`@`), with some questions and strategies outlined below.

Additional comments are made on individual lines of code, but often apply throughout

DESCRIPTION

- Excellent

NEWS

- is this format parsed by utils::news() ?

vignettes

- this is an excellent overview
- Modstrings.Rmd:119 (suggestion) use simpler `GRanges("1:5", mod="m7g")` ?

R 

- AllClasses.R:1 I don't understand the role of this file

- AllGenerics.R:24 add a `signature=` argument to restrict dispatch to arguments for which dispatch is required, e.g., `x` only, to simplify reasoning about multiple dispatch.

- Modstrings-ModString.R:170 (suggestion) the extreme indentation makes lines hard to parse, visually, and makes for very short space for lines. Consider a formatting convention that allows for consistent, modest indentation, e.g.,

    ```
    setReplaceMethod(
      "seqtype", "ModString",
      function(x, value)
    {
      ans_class <- paste(value, "String", sep="")
      if(is(x,ans_class)){
        return(x)
      }
      ans_seq <- .Call2(
        "new_CHARACTER_from_XString",
        x, 
        NULL,
        PACKAGE = "Biostrings")
      ans_seq <- .convert_one_byte_codes_to_originating_base(
        ans_seq,
        modscodec(seqtype(x)))
      do.call(ans_class,list(ans_seq))
    })
    ```

- Modstrings-ModString.R:192 Sorry for my ignorance, but this direct slot access looks very fragile to me; is there an API that can be used in XVector to access this? Minimally, follow the suggestion for `.Call()` below, to isolate the fragile portions of your code.

- Modstrings-ModString.R:208 consider `paste0()` when `sep = ""`.

- Modstrings-ModString.R:176 directly invoking `.Call()` from another package is very fragile; are there R-level APIs to these functions? Minimally, can I suggest consolidating all calls into a single 'interface' file with entries like

    ```
    .Call_new_Character_from_XString <- function(...)
        .Call2("new_CHARACTER_from_XString", ...)
        
    .Call_...
    ```
    
  to isolate the fragile portions of the code?

- Modstrings-ModString.R:282 this code looks quite unusual; why is it idfferetn from simply `seqtype(ans) <- seqtype` ?

- Modstrings-ModStringCodec-data.R: is this file appropriate, here, or perhaps under, e.g., `inst/script` or similar?

- Modstrings-ModStringCodec.R:35 it is unusual to have code like this evaluated at install time; perhaps these are pre-computed data objects stored as `inst/data/x.rds`?

- Modstrings-ModStringCodec.R:58 `initialize` methods are very tricky to get correct; generally our 'best practice' is to write a plain-old constructor that forms user-facing arguments into their necessary implementation format, then calls `new()`. Problems to consider here are (a) line 84 each assignment copies the entire object and (b) line 84 direct slot access is discouraged, one would rather accessors for each slot; for these objects it seems like 'setters' are not likely required...; (c) line 60 `initialize()` needs to act as a copy constructor for class extensions, but the current signature does not allow that.

- Modstrings-ModStringCodec.R:171 generally, use `message()` to communicate with the user, rather than `print()` or `cat()`. `message()`, like `stop()` and `warning()` sends output to the error stream rather than the output stream.

- Modstrings-ModStringCodec.R:233 calls to compiler::cmpfun are not necessary, package code is compiled at installation time by default.

- Modstrings-ModStringSet-io.R:41 fragile use of non-exported functions; approach like for `.Call()`, above, e.g., isolate to file and define `.close_filexp_list <- Biostrings:::.close_filexp_list`

- Modstrings-ModStringSet-io.R: it's difficult to parse this code on-the-fly; to what extent is it 'vectorized' to process many fasta / fastq records per R command?

- Modstrings-ModStringSet-io.R:585 is it necessary to use character strings / do.call? Instead along the lines of

    ```
    conFUN <- switch(compress, no = file, gzip = gzfile, ...)
    con <- conFUN(description=filepath2, open = ifelse(append, "ab", "wb", ...)
    ```
    
- Modstrings-matchLRPattern.R:21 this ['copy-and-append' pattern][1] is very inefficient, either 'pre-allocate and fill', or if elements are of variable size pre-allocate a list, fill with elements, and unlist.

[1]: http://bioconductor.org/developers/how-to/efficient-code/

**Comment by FelixErnst** [comment]
Thanks for the review.

The use of unexported functions, the C calls and direct slot access are all a results of the patterns outlined by the Biostrings package. Currently I don't see a clear way around them, without changes to a number of core BioC packages. Since the Modstrings package is designed to be an extension of the Biostrings package and I intend to follow the development of the Biostrings package closely, I don't see this as a problem.

NEWS
- yes I tested this and the examples on the bioconductor website look the same.

R
- AllClasses.R: Obsolete. deleted

- AllGenerics.R: done. Forgot about after finishing the functions

- Modstrings-ModString.R:170 thanks for the suggestions.

- Modstrings-ModString.R:192 This follows the use case outlined in the Biostrings package. In addition, there are no accessors for the `XRaw` or `XVector` class, which would avoid accessing the slots directly

- Modstrings-ModString.R:208 done

- Modstrings-ModString.R:176 All of the `.Call()` calls are carried over from the Biostrings package, which are used there within a wrapper. I wanted to closely resemble the code outlined in Biostrings package to be able to follow development of that package as close as possible. Therefore, I modified the wrapper to the needs of the Modstrings package, which was necessary because of the special encoding situation. So all the `.Call()` calls are already wrapped as much as they can be. I could wrap each of them as you suggested, but that would just add a function call and provide no additional benefit from my point of view.

-  Modstrings-ModString.R:282 This basically short hand for
```
x <- ModDNAString("AGCT")
x <- as(x,"DNAString") # 1
x <- as(x,"RNAString") # 2
x <- as(x,"ModRNAString") # 3
```
without the additional route around the `setAs` function. I think this is the shorter version. I admit that this is a fringe case, but I thought that some one might want convert a ModRNAString into a ModDNAString for whatever reason. In #1 all positions are switch back to the originating base. In #2 T/U conversion happens as designed in Biostrings package. The XString class is converted back into a ModString class in #3

- Modstrings-ModStringCodec.R:35 This is necessary, since characters are encoded differently on Linux and Windows. With this implementation the characters are encoded correctly either way. Every other option needs additional (and probably painfully) complicated logic.

- Modstrings-ModStringCodec.R:58 redone the constructor. Is this what you ment with "act as a copy constructor" not having an initialize function because additional slots cannot be added easily? 

- Modstrings-ModStringCodec.R:233 removed

- Modstrings-ModStringSet-io.R:41 I am not sure, what the benefit is of calling the non exported function this way. Can you explain, why wrapping it in a simple one liner makes sense? If the function name/behavior changes in the Biostrings package, I have to change it anyway and there is only one call to this specific function.

- Modstrings-ModStringSet-io.R reading is vectorized with a `mapply` call (line 169/320) to the connections. Each file is then read in 10000 lines at a time. Writing is available only per StringSet at a time as outlined by the Biostrings package (I don't want to deviate from the expected functionality).

- Modstrings-ModStringSet-io.R:585 done.

- Modstrings-matchLRPattern.R:21 this is basically carried over 1:1 from the Biostrings package, since I just needed to change the last function call from `unsafe.newXStringViews` to `unsafe.newModStringViews`. I can set that this is not the most efficient code style, but the functionality from the Biostrings package makes it necessary, or at least I haven't come up with an alternative solution. From my point of view, it would require some change in the way Views are constructed, which leads to the IRanges package as well, since basically all Views are constructed using the non-exported `new_Views` function from the IRanges package. Maybe a dynamic classname generator for the View class name might make this function in the Modstrings package obsolete, but this requires changes in the Biostrings package.

**Comment by mtmorgan** [comment]
Without having yet gone through the details of your response, the use of non exported symbols, .Call, and `@` violates the separation of interface and implementation that is central to maintainable, scalable software. That you are a conscientious maintainer now has little relevance in the long lifespan of your package. Conversely, the maintainers of the core packages are also conscientious and spend many hours considering the consequences of changes they make; changes to the implementation have consequences that are limited to the package itself, rather than to the broad ecosystem of packages, and hence are much more manageable, but implementations such as yours force them to spend a great deal of time worrying about packages that violate the API, so in the end it is very counterproductive to do this.

Normally, with a package like yours that is so invasive, I would simply say 'sorry, that is not acceptable'. The suggestion to encapsulate and localize the violations serves as an alternative to 'grep'ing your source code to identify all the subtle ways in which you violate the interface. For instance I can easily find that there 22 .Call statements in your code, but I currently have to do additional work to identify the exact number of distinct entry points there are. My intention is to then look at the specific calls that you make, and ask whether they are necessary and if so what changes to Biostrings / etc. would allow use of a public API.

It seems for instance that your call to `.new_CHARACTER_from_XString` requires that the lookup table (second argument) is NULL. In the public `Biostrings::as.character,XString-method`, this occurs when `Biostrings:::xs_dec_lkup` returns NULL, which in turn requires `Biostrings:::xscodec()` to return NULL, which requires `seqtype()` to return something other than `DNA` or `RNA`, which it already does
```
> seqtype(ModDNAString())
[1] "ModDNA"
```
So the public `as.character,XString-method` simply works without needing `.Call()` in your package and the implementation of `as.character,ModString-method` would appear to be something like
```
setMethod(
    "as.character", "ModString",
    function(x)
{
    ans <- callNextMethod()
    .convert_one_byte_codes_to_letters(ans, modscodec(seqtype(x)))
})
```

It is interesting that in your current code there are several calls to `new_CHARACTER_from_XString`, but only a single call in Biostrings. The single call in Biostrings means that there is only one point where the consequences of changes to the C-level API need to be considered -- an example of the advantage to separation of interface and implementation that we are discussing at a 'higher level' in your package. 

The above analysis took me about an hour to work through; I do not intend to do this for each use of `.Call()`, but instead viewed consolidation of violations into a single 'point of entry' as a first step in simplifying this analysis, e.g., knowing that there are not 22 .Call statements to analyze but a smaller number.

**Comment by FelixErnst** [comment]
Thanks for the clarification. I understand that the additional encapsulation would bundle all the the .Call statements. I put them into a separate file to make the origin of these functions clear.

Regarding the number of .Call statements:
- `new_CHARACTER_from_XString` was used two times, but there is a similar function called `new_CHARACTER_from_XStringSet`. The Biostrings package uses both of the functions one time, but I had to used an additional time, since `seqtype` on a `ModString` object is only used to converted modified nucleotides to the originating base. For this I cannot call `as.character`. However, for `as.character` I used your suggestions and handled it now with `callNextMethod`. Thanks for catching this.
- Each other .Call statement was used once, except `XString_replace_letter_at` and `XStringSet_vmatch_pattern_at`. Now they are used once in the encapsulated form.

Regarding the use of direct slot access:
- Direct slot access for `XVector` objects: I implemented the accessors for the `ModString` class only.
- I removed the direct slot replacement for `MaskedModStrings`. They now use the replace methods from the Biostrings package.

Regarding the use of non-exported functions:
- All functions used, which are from another package and not exported by this package, are encapsulated in separate files as well. I have two remaining functions reported as used with `:::`. However, I cannot find them anywhere in my code and I am also not aware ever having used them. Would you have a suggestion, what might be going on?

Regarding the use of message:
- The one instance, where I use `print` instead of `message`/`error` is due to the fact, that both functions change the encoding of the message. This leads to non usable output due to the special characters.

**Comment by mtmorgan** [comment]
Since the direct call to `"new_CHARACTER_from_XString"` can be replaced by a simple `as.character()`, shouldn't your code just use `as.character()` instead?

**Comment by FelixErnst** [comment]
I think you are referring to the replace `seqtype` function, correct? `"new_CHARACTER_from_XString"` can not be removed from there. The replace `seqtype` function is used in the `Biostrings` package to handle type conversions and every `XString` has to have one. In case of a `ModString` there are two options.

1. `seqtype` gets called with matching arguments, lets say `"ModRNA"` as `value` and a `ModRNAString` as `x`. `x` gets returned immediately.
2. `seqtype` gets called with non-matching arguments. Since `seqtype` is usually called from coercion function, the most likely scenario is, that `seqtype` is called with a `ModRNAString` as `x` and `RNA` as value. Here the task is to revert all modified nucleotides to there originating base. `as.character` could be used, but in this case all modified nucleotides would need to be converted twice. First, in `as.character` from their byte value to the special character and, secondly, the special character to the originating base. I opted to omit the unnecessary conversion and go straight from the byte value to the originating base.
(The combination `ModRNAString` as `x` and `DNA` as value leads to an error, because U is not a letter in the DNA alphabet).

Making `Biostrings:::xscodes()` a (public? It does not necessarily need to be exported, would it?) generic would make some calls simpler, but would probably not eliminate all `.Call.` calls. For example in `_letter_frequency` the names of the codes have to be set to the special characters, this is done via `modscodes`, which accepts the additional argument `lettersOnly`. If set to `TRUE`, this returns the special characters instead of the byte value. Having a public generic `xscodes` would allow the functionality of `modscodes` to be incorporated into the specific `xscodes` function. But reformating the names, would still need to be done....... but this could be done afterwards. 

**You are right**, this would allow the higher level methods for `XString_letter_frequency`,  `XStringSet_letter_frequency`,  `XString_letterFrequencyInSlidingView`, `XStringSet_letterFrequency` and `XStringSet_consensus_matrix` from the `Biostrings` package to be used.

For the some of the remaining `.Call` calls to be removed and higher level functions to be usable directly, the following function would need to generics as well:

- `.normargSubject`
- `.normargPattern`
- `.normargFixed`
- `.normarg_value1`
- `.normarg_value2`

In addition, `.normargSubject`, would need to be usable in `.XString.matchPattern` (Currently only a `if` call is used, mimicking a minimal version of  `.normargSubject`: line 127 in matchPattern.R). The same goes for `.XStringSet.vmatchPattern`.

I don't know, what @hpages thinks about these potential changes. Having the generics would makes thinks more tidy and more interactable. How fast this can be reworked and if it all can play out as we outlined here for the `Biostrings` package, I think, I need to leave for both of you to judge (Probably more people involved?). I don't want to demand these changes, since I am grateful, that I even can play with the modified nucleotides on this scale. I know you have enough on your plates.

I currently don't see an option to remove the calls to the following function:
- `XString_replace_letter_at`: First, I need it for the `modifyNucleotide` function. Secondly, in `replaceNucleotideAt` the input needs to be an `XString`, then I need to convert letters to byte value, and the `.Call2` gets called. Since I need the middle part, I don't see how this could be simplified, so that I can use the higher level function.
- `fasta_index`, because I cannot use the other writing functions, since they cannot handle the special characters.

**Comment by FelixErnst** [comment]
I retraced my notes and the structure a bit deeper for the other ~~five~~ two functions, which hinder the use of the high level calls in the Biostrings package.

- `normargSubject`: This is not needed, anymore. It was during development, but I simplified somethings and it is not longer needed. I made a mistake here.
- `normargPattern`: With this function the problem here is, that the internal constructor `XString` is called (lowlevel-matching.R: line 34). This function does not do the letter to byte conversion, which I need. Therefore I need a custom `.normargPattern`, and therefore need the other custom functions. If this would be changed to `pattern <- try(do.call(xsbaseclass(subject),list(pattern))` in Biostrings, this would work out of the box. In addition, these changes would also need to be applied to `lowlevel-matching.R line 434` for the `hasLetterAt` function.
- `normargFixed`: This is already imported from `Biostrings`. I scratched out in the comment above.
- `normarg_value1`: See `.normargSubject`
- `normarg_value2`: See `.normargSubject`
- `unsafe.newXStringViews` (`matchPattern.R` line 145 and 173 and `matchLRPattern.R` line 40 and 87): dynamic class name generation would be need in `unsafe.newXStringViews`, since the ModStringViews class needs to implemented (`XString.read` in `letters,XStringViews-method`, the comparison functions and the `show` function). Alternatively, `unsafe.newXStringViews` could be exchanged to the `Views` function in the `Biostrings` package (at least in the four lines I outlined above). I couldn't find any hint, why this "unsafe" constructor would be needed. However, `unsafe.newXStringViews` is used at several places, so I don't what problems this solves at all these places.

So the except for `xscodes` not additional generics would be need, if the constructor calls wouldn't use the internal constructor and `Views` could be constructed more dynamically from the . That would drop the use of `.Call2` to 6 (`new_CHARACTER_from_XString`,  `new_XString_from_CHARACTER`, `new_CHARACTER_from_XStringSet`, `new_XStringSet_from_CHARACTER`, `XString_replace_letter_at` and `fasta_index`) and would remove half of the imported non-exported functions from the `Biostrings` package.

**Comment by FelixErnst** [comment]
The warnings and timeout are related to connection issues on the Linux system. Windows and OSX are find.

**Comment by FelixErnst** [comment]
On the Modstrings repo there is now an additional branch to match the suggested changes in the Biostrings package.

https://github.com/FelixErnst/Modstrings/tree/Biostrings-2.51.3

If the changes are incorporated into Biostrings and have propagate to the SPB, I can merge the changes to master.

**Comment by FelixErnst** [comment]
So I did and additional iteration of changes in Biostrings package. This proved to be quite helpful, with all the constructurs of the `ModString`, `ModStringSet` and `ModStringSetList` constructors.

I am ready to proceed with the next iteration in the review. I don't know how long it will take until the changes in `Biostrings ` package take effect on the SPB. I will merge the changes as soon as I can, In the meantime they will remain in the branch mentioned above.

**Comment by lshep** [comment]
Once the pull request changes are accepted and merged into the Biostrings repository, they will need to propagate to the [daily builder](http://bioconductor.org/checkResults/3.9/bioc-LATEST/), once there it will be available on the SPB.   The SPB doesn't have the ability to automatically build other branches besides the master branch.  

The timestamp issue will be corrected when we update R-devel ( which we will do this week).   The biocViews ERROR we should have resolved soon as well but can be ignore as this is on our end.

**Comment by FelixErnst** [comment]
Hi @mtmorgan 
with the efforts of @hpages the Modstrings got a bit simpler. Ready for the next round of the review.

**Comment by FelixErnst** [comment]
@lshep Hi, I get warning about the Authors@R field. I didn't change anything on that field. Any idea what I can do?

**Comment by lshep** [comment]
Sorry.  There was a new check added but there must be a bug somewhere. Please ignore for now and I'll look into this right away

**Comment by FelixErnst** [comment]
@mtmorgan @hpages  Thank you very much for the review and the work you put in for making the interaction with the Biostrings package more easy! It is much appreciated.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/FelixErnst.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("Modstrings")`. The package 'landing page' will be created at

https://bioconductor.org/packages/Modstrings

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
